Add Autofocus, IR lamp, and Wiper switches in ONVIF

partofthething opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This adds switch entities to allow users to send ONVIF AuxiliaryCommands and ImagingSettings to cameras that turn on and off wipers, autofocus, and IR lamps.

For example, with this, automations can be built on top of weather data that run the wiper if it's raining before taking a scheduled snapshot.



## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/25370

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

partofthething committed:
Add support for AuxiliaryCommands in ONVIF.

This enables users to send commands that turn on wipers,
heaters, etc.

dgomes reviewed (commented):
Why call it Command if it is a Service ?

partofthething commented:
> Why call it Command if it is a Service ?

It's called an AuxiliaryCommand in ONVIF but a service in Home Assistant. 

![image](https://user-images.githubusercontent.com/5360219/209976679-458d4829-811b-42d0-8f28-0a1d02fdc394.png)

source: https://www.onvif.org/specs/srv/ptz/ONVIF-PTZ-Service-Spec-v221.pdf

partofthething committed:
Add ONVIF Imaging Settings service.

frenck commented:
I don't understand why these are services and not just entities?

partofthething committed:
Turn AutoFocus toggle and wiper control into switch entities.

This exposes more user-friendly entities for controlling
these.

It does leave the underlying services exposed in case power-users
have other camera features they want to use that are not generally
supported.

partofthething commented:
> I don't understand why these are services and not just entities?

Great point. After your pointers in the documentation PR I have attempted to make two switch entities in my latest commit. If this approach is more in-line with your expectations, I can go ahead and add the other imaging settings as switch and number entities in the same way.

partofthething committed:
Fix typing.

partofthething committed:
Fix mistake in turn off.

GrumpyMeow commented:
Almost there guys! Only add a unit-test to get the codecoverage in order...

partofthething commented:
> Almost there guys! Only add a unit-test to get the codecoverage in order...

Ok I'll try to get it there. I looked over the existing tests and started trying to make the tests, but frankly got a little lost in how exactly to set up the test fixtures for this situation. I'll figure it out though. sorry for delay.

partofthething committed:
Update tests for onvif switch entities.

Changed how profile and init gets done to simplify
test fixtures.

partofthething committed:
Ran linters and formatters and added more tests.

MartinHjelmare commented on homeassistant/components/onvif/__init__.py:
```diff
@@ -60,8 +63,10 @@ async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+        platforms += [Platform.BINARY_SENSOR, Platform.SENSOR]
         await device.events.async_stop()
+    if device.capabilities.imaging:
+        platforms += [Platform.SENSOR]
```
Should this be a switch platform as that's what we added for imaging when setting up the config entry?

```suggestion
        platforms += [Platform.SWITCH]
```

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -82,11 +86,26 @@ async def async_setup_entry(
-    async_add_entities(
-        [ONVIFCameraEntity(device, profile) for profile in device.profiles]
-    )
+    entities = [ONVIFCameraEntity(device, profile) for profile in device.profiles]
```
This change doesn't look needed.

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -221,3 +240,23 @@ async def async_perform_ptz(
+
+    async def async_run_aux_command(
+        self,
+        cmd,
```
```suggestion
        cmd: str,
```

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -221,3 +240,23 @@ async def async_perform_ptz(
+
+    async def async_set_imaging_settings(
+        self,
+        settings,
```
```suggestion
        settings: dict,
```

MartinHjelmare commented on homeassistant/components/onvif/device.py:
```diff
@@ -482,6 +493,63 @@ async def async_perform_ptz(
+    async def async_run_aux_command(
+        self,
+        profile: Profile,
+        cmd,
```
```suggestion
        cmd: str,
```

MartinHjelmare commented on homeassistant/components/onvif/device.py:
```diff
@@ -482,6 +493,63 @@ async def async_perform_ptz(
+    async def async_set_imaging_settings(
+        self,
+        profile: Profile,
+        settings,
```
```suggestion
        settings: dict,
```

MartinHjelmare commented on homeassistant/components/onvif/device.py:
```diff
@@ -482,6 +493,63 @@ async def async_perform_ptz(
+        self,
+        profile: Profile,
+        cmd,
+    ):
```
```suggestion
    ) -> None:
```

MartinHjelmare commented on homeassistant/components/onvif/device.py:
```diff
@@ -482,6 +493,63 @@ async def async_perform_ptz(
+        self,
+        profile: Profile,
+        settings,
+    ):
```
```suggestion
    ) -> None:
```

MartinHjelmare commented on homeassistant/components/onvif/switch.py:
```diff
@@ -0,0 +1,108 @@
+    _on_settings: dict[str, Any] = {}
+    _off_settings: dict[str, Any] = {}
+
+    async def async_turn_on(self, **kwargs) -> None:
```
```suggestion
    async def async_turn_on(self, **kwargs: Any) -> None:
```

MartinHjelmare commented on homeassistant/components/onvif/switch.py:
```diff
@@ -0,0 +1,108 @@
+        profile = self.device.profiles[0]
+        await self.device.async_set_imaging_settings(profile, self._on_settings)
+
+    async def async_turn_off(self, **kwargs) -> None:
```
```suggestion
    async def async_turn_off(self, **kwargs: Any) -> None:
```

MartinHjelmare commented on homeassistant/components/onvif/switch.py:
```diff
@@ -0,0 +1,108 @@
+        await self.device.async_set_imaging_settings(profile, self._off_settings)
+
+
+class ONVIFAuxSwitch(ONVIFBaseEntity, SwitchEntity):
```
I suggest using entity descriptions and a single switch entity class instead of multiple classes. Pass the `device` and `profile` parameters to the turn on / off lambdas of the description.

Example:

https://github.com/home-assistant/core/blob/cbcff6435f4efdfd5e92aa5394f776b990666726/homeassistant/components/adguard/switch.py#L24-L48

https://github.com/home-assistant/core/blob/cbcff6435f4efdfd5e92aa5394f776b990666726/homeassistant/components/adguard/switch.py#L113-L145

partofthething replied:
Good idea. Ok I think I have accomplished this in the latest commit.

MartinHjelmare commented on homeassistant/components/onvif/services.yaml:
```diff
@@ -76,3 +76,33 @@ ptz:
+
+aux:
+  name: AUX
+  description: If your ONVIF camera supports PTZ auxiliary commands such as turning wiper or heater on, you will be able to send them.
```
Are we still going to implement the custom services when we add the switches?

partofthething replied:
I think it's wise to keep the service for now because of the massive breadth of the ONVIF protocol covering thousands of different camera models. The entities expose some common/expected functionality, but if someone wants to do something that only their device supports, the services will allow that. I think most future development will focus on the user-friendly entities.

MartinHjelmare replied:
If someone wants a new feature we can add a new switch (description) or similar. I suggest we don't add the services until explicitly required.

partofthething replied:
Ok I removed them.

partofthething committed:
Update homeassistant/components/onvif/__init__.py

Fix incorrect platform.

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

partofthething committed:
Add some type hints from @MartinHjelmare's review.

partofthething committed:
Switch to EntityDescriptions

Suggested by MartinHjelmare in review.

partofthething committed:
Attempting to fix type errors.

partofthething commented:
Anyone know how to fix those mypy errors? I struggled for a while but couldn't figure it out.

```
Error: homeassistant/components/onvif/switch.py:93: error: "SwitchEntityDescription" has no attribute "turn_on_fn"  [attr-defined]
Error: homeassistant/components/onvif/switch.py:94: error: "SwitchEntityDescription" has no attribute "turn_on_data"  [attr-defined]
Error: homeassistant/components/onvif/switch.py:101: error: "SwitchEntityDescription" has no attribute "turn_off_fn"  [attr-defined]
Error: homeassistant/components/onvif/switch.py:102: error: "SwitchEntityDescription" has no attribute "turn_off_data"  [attr-defined]
```

partofthething committed:
Add type hint

MartinHjelmare commented on homeassistant/components/onvif/switch.py:
```diff
@@ -13,6 +15,54 @@
+class ONVIFSwitchEntityDescriptionMixin:
+    """Mixin for required keys."""
+
+    turn_on_fn: Callable[[Any], Any]
```
```suggestion
    turn_on_fn: Callable[[ONVIFDevice], Callable[[Any], None]]
```

partofthething replied:
This appears to lead to:
```
switch.py:42: error: Argument "turn_on_fn" to "ONVIFSwitchEntityDescription" has incompatible type "Callable[[ONVIFDevice], Callable[[Profile, Dict[Any, Any]], Coroutine[Any, Any, None]]]"; expected "Callable[[ONVIFDevice], Callable[[Any], None]]"  [arg-type]
switch.py:42: error: Incompatible return value type (got "Callable[[Profile, Dict[Any, Any]], Coroutine[Any, Any, None]]", expected "Callable[[Any], None]")  [return-value]
switch.py:43: error: Argument "turn_off_fn" to "ONVIFSwitchEntityDescription" has incompatible type "Callable[[ONVIFDevice], Callable[[Profile, Dict[Any, Any]], Coroutine[Any, Any, None]]]"; expected "Callable[[ONVIFDevice], Callable[[Any], None]]"  [arg-type]
```
during mypy. I struggled for a long time to get mypy happy on this and ended up on the general one above. Can keep trying.

partofthething replied:
ah nm I think I got it.

MartinHjelmare commented on homeassistant/components/onvif/switch.py:
```diff
@@ -21,88 +71,35 @@ async def async_setup_entry(
+    ) -> None:
+        """Initialize the switch."""
+        super().__init__(device)
+        self._attr_name = f"{self.device.name} {description.name}"
```
Please set the class attribute `_attr_has_entity_name` to `True` instead and remove this line.

https://developers.home-assistant.io/docs/core/entity#entity-naming

MartinHjelmare commented on homeassistant/components/onvif/switch.py:
```diff
@@ -21,88 +71,35 @@ async def async_setup_entry(
-        self._attr_is_on = True
-        profile = self.device.profiles[0]
-        await self.device.async_set_imaging_settings(profile, self._on_settings)
+    async_add_entities([ONVIFSwitch(device, description) for description in SWITCHES])
```
```suggestion
    async_add_entities(ONVIFSwitch(device, description) for description in SWITCHES)
```

MartinHjelmare commented on homeassistant/components/onvif/switch.py:
```diff
@@ -13,6 +15,54 @@
+    ),
+    ONVIFSwitchEntityDescription(
+        key="ir_lamp",
+        name="IR Lamp",
```
```suggestion
        name="IR lamp",
```

MartinHjelmare commented on homeassistant/components/onvif/switch.py:
```diff
@@ -13,6 +15,54 @@
+SWITCHES: tuple[ONVIFSwitchEntityDescription, ...] = (
+    ONVIFSwitchEntityDescription(
+        key="autofocus",
+        name="AutoFocus",
```
```suggestion
        name="Autofocus",
```

partofthething committed:
Remove services, update typing, apply other suggestions.

partofthething committed:
Fix accidental removal of PTZ constant, and typing imports.

MartinHjelmare commented on homeassistant/components/onvif/camera.py:
```diff
@@ -221,3 +221,23 @@ async def async_perform_ptz(
             zoom,
         )
+
+    async def async_run_aux_command(
```
We don't need these additions to the camera platform now, right?

partofthething replied:
that's correct, thanks. Removing now.

MartinHjelmare commented on tests/components/onvif/test_switch.py:
```diff
@@ -0,0 +1,149 @@
+    )
+    await hass.async_block_till_done()
+
+    device.async_run_aux_command.assert_called_once()
```
Please assert the new state of the switch entity in the state machine after the service call.

partofthething replied:
good idea, thanks. done.

partofthething committed:
Remove camera methods related to imaging and aux.

These should have been removed when the related services
were removed in an earlier commit.

partofthething committed:
Test state machine changes for new switch entities.

partofthething committed:
Fix import order

MartinHjelmare reviewed (approved):
Looks good!

MartinHjelmare commented on tests/components/onvif/test_switch.py:
```diff
@@ -0,0 +1,161 @@
+"""Test button of ONVIF integration."""
```
```suggestion
"""Test switch platform of ONVIF integration."""
```

MartinHjelmare committed:
Adjust docstring
